### PR TITLE
JSON schema pattern properties support in sources

### PR DESCRIPTION
**Pull request by raphaelsolarski**

### CODE REVIEW

File: docs/integration/DataTypingAndSchemasHandling.md [hunk 1]
```diff
@@ -202,8 +202,17 @@ These properties will be not validated by the Designer, because on during scenar
 `Typing Information` not on real value. Validation will be still done at runtime.
 
 #### Pattern properties
-Pattern properties are supported only in sinks. Using patternProperties with sources can produce runtime errors.
+WhPattern properties are supported in sources and sinks.
```
Comment (gadomsky) [inline docs/integration/DataTypingAndSchemasHandling.md:205]: typo?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 2]
```diff
@@ -171,23 +174,32 @@ object SwaggerArray {
...
+  private implicit val patternKeyEncoder: KeyEncoder[Pattern] = KeyEncoder[String].contramap(_.pattern())
+
   private[swagger] def apply(schema: Schema[Object], swaggerRefSchemas: SwaggerRefSchemas, usedRefs: Set[String]): SwaggerTyped = {
     val properties = Option(schema.getProperties).map(_.asScala.toMap.mapValuesNow(SwaggerTyped(_, swaggerRefSchemas, usedRefs)).toMap).getOrElse(Map())
 
     if (properties.isEmpty) {
       schema.getAdditionalProperties match {
-        case a: Schema[_] => SwaggerMap(Some(SwaggerTyped(a, swaggerRefSchemas, usedRefs)))
+        case a: Schema[_] if schema.getPatternProperties == null || schema.getPatternProperties.isEmpty => SwaggerMap(Some(SwaggerTyped(a, swaggerRefSchemas, usedRefs)))
         case b if b == false => new SwaggerObject(Map.empty, AdditionalPropertiesDisabled)
```
Comment (gadomsky) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:186]: here in `SwaggerObject`, why don't you handle `patternProperties` ?

Comment (raphaelsolarski) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala]: fixed + test case added

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 3]
```diff
@@ -164,30 +163,49 @@ object SwaggerTyped {
...
+    import pl.touk.nussknacker.engine.util.Implicits.RichScalaMap
+    def typedStringKeyMap(valueType: TypingResult) = {
+      Typed.genericTypeClass(classOf[util.Map[_, _]], List(Typed[PropertyName], valueType))
+    }
+    if (elementType.isEmpty) {
+      val patternPropertiesTypesSet = patternProperties.values.map(typingResult).toSet
+      additionalProperties match {
+        case AdditionalPropertiesDisabled if patternPropertiesTypesSet.isEmpty => TypedObjectTypingResult(List.empty)
+        case AdditionalPropertiesDisabled => typedStringKeyMap(Typed(patternPropertiesTypesSet))
+        case AdditionalPropertiesWithoutType => typedStringKeyMap(Unknown)
```
Comment (gadomsky) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:177]: here, you dont want to handle patternProperties?

Comment (raphaelsolarski) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:177]: nope, but thanks for ensuring

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 4]
```diff
@@ -68,9 +70,9 @@ object SwaggerEnum {
 
 case class SwaggerArray(elementType: SwaggerTyped) extends SwaggerTyped
 
-case class SwaggerObject(elementType: Map[PropertyName, SwaggerTyped], additionalProperties: AdditionalProperties = AdditionalPropertiesWithoutType) extends SwaggerTyped
-
-case class SwaggerMap(valuesType: Option[SwaggerTyped]) extends SwaggerTyped
+@JsonCodec case class SwaggerObject(elementType: Map[PropertyName, SwaggerTyped],
+                                    additionalProperties: AdditionalProperties = AdditionalPropertiesWithoutType,
+                                    patternProperties: Map[Pattern, SwaggerTyped] = Map.empty) extends SwaggerTyped
```
[the remark this answers is not in the document]
Comment (raphaelsolarski) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:75]: good point

File: docs/integration/DataTypingAndSchemasHandling.md [hunk 5]
```diff
@@ -202,8 +202,17 @@ These properties will be not validated by the Designer, because on during scenar
...
 
+##### Sources
+Object (also nested) in source schema will be represented during scenario authoring as:
+* Map - when there is no property defined in "properties" field
+  * if only additional properties are defined then map values will be typed to according to schema in additionalProperty field
+  * if both additional and pattern properties are defined then values will be typed to `Unknown` and can be casted to exact type by end user
```
[the remark this answers is not in the document]
Comment (raphaelsolarski) [reply docs/integration/DataTypingAndSchemasHandling.md]: fixed

File: docs/integration/DataTypingAndSchemasHandling.md [hunk 6]
```diff
@@ -202,8 +202,16 @@ These properties will be not validated by the Designer, because on during scenar
...
 #### Pattern properties
-Pattern properties are supported only in sinks. Using patternProperties with sources can produce runtime errors.
 
+##### Sources
+Object (also nested) in source schema will be represented during scenario authoring as:
+* Map - when there is no property defined in "properties" field
+  * if only additional properties are defined then map values will be typed to according to schema in additionalProperty field
+  * if both additional and pattern properties are defined then values will be typed to `Unknown` and can be casted to exact type by end user
+* Record otherwise
+  * pattern and additionalProperties can then be accessed using `record["patternOrAdditionalPropertyName"]` syntax but only if `pl.touk.nussknacker.engine.api.process.ExpressionConfig.dynamicPropertyAccessAllowed` is enabled
```
[the remark this answers is not in the document]
Comment (raphaelsolarski) [reply docs/integration/DataTypingAndSchemasHandling.md]: fixed

**Review by arkadius — APPROVED** [review]
LGTM

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 7]
```diff
@@ -87,6 +86,8 @@ sealed trait SwaggerUnknownFallback extends SwaggerTyped
 
 case object SwaggerRecursiveSchema extends SwaggerUnknownFallback
 
+case object SwaggerAny extends SwaggerUnknownFallback
```
Comment (arkadius) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:89]: One slight detail. IMO after this change `SwaggerUnknownFallback` has no sence. It was fallback fall situation when recursive schema was detected. Maybe we should have `SwaggerAny` trait and two objects showing in which case it was used: `SwaggerRecursiveSchema` and `SwaggerAny` (the same name, but object). WDYT?

**Review by arkadius — APPROVED** [review]
Nice

**Pull request merged**